packages/palette/src/elements/MultiSelect/MultiSelect.tsx

```diff
@@ -0,0 +1,172 @@
+import { themeGet } from "@styled-system/theme-get"
+import React, { useEffect, useState } from "react"
+import styled from "styled-components"
+import { DROP_SHADOW } from "../../helpers"
+import { useClickOutside, usePosition } from "../../utils"
+import { useUpdateEffect } from "../../utils/useUpdateEffect"
+import { Box, BoxProps } from "../Box"
+import { Checkbox } from "../Checkbox"
+import { Clickable } from "../Clickable"
+import { caretMixin, Option } from "../Select"
+import { Sup } from "../Sup"
+import { Text } from "../Text"
+
+export interface MultiSelectProps extends BoxProps {
+  title: React.ReactNode
+  options: Option[]
+  onSelect?: (selection: Option[]) => void
+}
+
+export const MultiSelect: React.FC<MultiSelectProps> = ({
+  title,
+  options,
+  onSelect,
+  ...rest
+}) => {
+  const [visible, setVisible] = useState(false)
+  const [selection, setSelection] = useState<Option[]>([])
+  const [width, setWidth] = useState(0)
+
+  // Yields focus back and forth between popover and anchor
+  useUpdateEffect(() => {
+    if (visible && tooltipRef.current) {
+      tooltipRef.current.focus()
+      return
+    }
+
+    if (!anchorRef.current) return
+    anchorRef.current.focus()
+  }, [visible])
+
+  const onVisible = () => setVisible(true)
+  const onHide = () => setVisible(false)
+
+  useEffect(() => {
+    const handleKeydown = (event: KeyboardEvent) => {
+      if (event.key === "Escape") {
+        onHide()
+      }
+    }
+
+    document.addEventListener("keydown", handleKeydown)
+    return () => {
+      document.removeEventListener("keydown", handleKeydown)
+    }
+  }, [])
+
+  useEffect(() => {
+    if (!anchorRef.current) return
+
+    const handleResize = () => {
+      setWidth(anchorRef.current.offsetWidth)
+    }
+
+    handleResize()
+
+    window.addEventListener("resize", handleResize)
+    return () => {
+      window.removeEventListener("keydown", handleResize)
+    }
+  }, [])
+
+  useUpdateEffect(() => {
+    onSelect?.(selection)
+  }, [onSelect, selection])
```

Comment:
I guess we want to run `onSelect?.(selection)` inside of the checkbox's `onSelect` callback. Updating the value with `useUpdateEffect` calls `onSelect` every time the parent component rerenders if it's used without `useRef` or `useCallback`.

Reply:
How's that the case? It should only call it if either `onSelect` or `selection` change.

Reply:
This will give a `Warning: Maximum update depth exceeded. `:

```
import { MultiSelect } from "@artsy/palette"
import React, { useReducer } from "react"

const sizeOptions = [
  { text: "Small (under 40cm)", value: "SMALL" },
  { text: "Medium (40cm - 100cm)", value: "MEDIUM" },
  { text: "Large (over 100cm)", value: "LARGE" },
]

export const PriceDatabaseSearch = () => {
  const [filters, setFilter] = useReducer(
    (filters, newFilter) => ({ ...filters, ...newFilter }),
    {}
  )

  return (
    <MultiSelect
      options={sizeOptions}
      title="Size"
      onSelect={sizes => {
        setFilter({ sizes: sizes.map(size => size.value) })
      }}
    />
  )
}
```

Reply:
Ah, thanks for the example. Should be fixed now. Give it another whirl once the canary re-publishes.

Reply:
That works - thanks for the fix!